### PR TITLE
Iso-duration

### PR DESCRIPTION
**Pull request by constantinius**
Proper ISO 8601 duration encoding for temporal dimension steps

### CODE REVIEW

**Review by gadomski — COMMENTED** [review]
Marking as draft until https://github.com/stactools-packages/datacube/pull/10 is merged, as this includes the diff from that.

**Review by gadomski — COMMENTED** [review]
Is https://pypi.org/project/isoduration/ usable? If there's an off-the-shelf package that we can lean on, that'd be preferable to a re-implementation IMO.

I ran into the same issue in **noaa-cdr**: https://github.com/stactools-packages/noaa-cdr/blob/db4ebdc633a2cb1f27874b039edcbe761b81b214/src/stactools/noaa_cdr/time.py#L191-L215. There, I knew I had a limited subset of cases to handle, so I _did_ re-implement, but since this is supposed to be a general-use library, we should avoid re-implementation if possible IMO.

**Comment by constantinius** [comment]
After some investigation I can say: unfortunately `isoduration` does not provide the arithmetic operations on their central `Duration` class that we need in this case.

The other option `pandas.Timedelta` is quite a heavy dependency and also does not cover all the bases.

For temporal dimensions we have a single conversion from `timedelta` to an ISO duration, which I hope is covered by the function. I would love to go ahead with this.

File: src/stactools/datacube/stac.py [hunk 1]
```diff
@@ -76,6 +76,32 @@ def get_time_offset_and_step(unit: str) -> Tuple[datetime, timedelta]:
     raise ValueError(f"Failed to parse time unit from '{unit}'")
 
 
+def iso_duration(td: timedelta) -> str:
```
Comment (gadomski) [inline src/stactools/datacube/stac.py:79]: Can you add a simple docstring describing the function?

File: src/stactools/datacube/stac.py [hunk 2]
```diff
@@ -76,6 +76,32 @@ def get_time_offset_and_step(unit: str) -> Tuple[datetime, timedelta]:
...
+    date_intervals = [
+        ('W', 604800),  # 60 * 60 * 24 * 7
+        ('D', 86400),    # 60 * 60 * 24
+    ]
```
Comment (gadomski) [inline src/stactools/datacube/stac.py]: Don't we need a `Y` and `M` as well?

File: src/stactools/datacube/stac.py [hunk 3]
```diff
@@ -76,6 +76,32 @@ def get_time_offset_and_step(unit: str) -> Tuple[datetime, timedelta]:
...
+        if value:
+            seconds -= value * count
+            result.append(f"{value}{name}")
+    result.append("T")
+    for name, count in time_intervals:
+        value = int(seconds // count)
+        if value:
+            seconds -= value * count
+            result.append(f"{value}{name}")
+    result.append(f"{seconds}S")
```
Comment (gadomski) [inline src/stactools/datacube/stac.py]: Omit if `seconds == 0`?

File: src/stactools/datacube/stac.py [hunk 4]
```diff
@@ -76,6 +76,32 @@ def get_time_offset_and_step(unit: str) -> Tuple[datetime, timedelta]:
...
+        ('M', 60),
+    ]
+    seconds = td.total_seconds()
+    result = ["P"]
+    for name, count in date_intervals:
+        value = int(seconds // count)
+        if value:
+            seconds -= value * count
+            result.append(f"{value}{name}")
+    result.append("T")
```
Comment (gadomski) [inline src/stactools/datacube/stac.py]: If we're e.g. exactly one day, we won't need the `T` right?

**Review by gadomski — CHANGES_REQUESTED** [review]
`iso_duration` deserves some unit tests, I think, to ensure we're handling some common cases properly. At minimum:

- One year
- One month
- One day
- One week
- One hour
- One second
- A duration that has everything
- A duration that doesn't have any hours, minutes, or seconds

That will help future developers understand the expected behavior of the function.

Question: do we need a reverse function, one that parses a string into a timedelta?

File: src/stactools/datacube/stac.py (hunk 4 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py]: That is correct. Will revise

File: src/stactools/datacube/stac.py (hunk 3 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py]: Yes, that is true. But if, for some reason, we have a zero-duration, it has to be present again

File: src/stactools/datacube/stac.py (hunk 2 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py]: Years and months are ambiguous, so we should not be using them

**Comment by constantinius** [comment]
> Question: do we need a reverse function, one that parses a string into a timedelta?

This has not yet come up. Currently we are building te timedeltas from netcdf/zarr metadata. They are usually expressed as somthing like `"days since 1990-01-01"`

File: src/stactools/datacube/stac.py (hunk 2 above)
Comment (gadomski) [reply src/stactools/datacube/stac.py]: Makes sense, we should just be careful to document that this is an "opinionated" function and explain the reasoning why.

File: src/stactools/datacube/stac.py (hunk 3 above)
Comment (gadomski) [reply src/stactools/datacube/stac.py]: That seems like a special case that can handled at the end (e.g. a `if len(result) == 1` or whatever).

Comment (constantinius) [reply src/stactools/datacube/stac.py]: Should be solved now

File: src/stactools/datacube/stac.py (hunk 2 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py]: Done

File: src/stactools/datacube/stac.py (hunk 1 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py:79]: Done

File: src/stactools/datacube/stac.py (hunk 4 above)
Comment (constantinius) [reply src/stactools/datacube/stac.py]: Done

**Comment by constantinius** [comment]
I added a couple of unit-tests

**Review by gadomski — APPROVED** [review]

**Comment by gadomski** [comment]
Appreciate the fixups, looks good!

**Review by gadomski — CHANGES_REQUESTED** [review]
@constantinius this is not passing CI so needs fixes before it can go in and we release.

**Review by gadomski — APPROVED** [review]

**Pull request merged**